### ISSUE TITLE
`Layout/SpaceBeforeSemicolon` -> `Layout/SpaceAfterComma` - Infinite loop detected with block local arguments

### ISSUE THREAD

**Issue by snutij** [rubocop/rubocop#12020]
Hello 👋  I have an issue, during formatting where two rules seems to create an infite loop: 

Given a file `foo.rb`:

```ruby
foo { |a, ; x|
}

```

Which is a valid syntax: 

```
ruby -c foo.rb
Syntax OK
```

--------

## Expected behavior

Should not trigger an error. And _I think_ it should remove comma, but not sure at all. 

## Actual behavior

`bin/rubocop -d -x --only Layout/SpaceAfterComma,Layout/SpaceBeforeSemicolon foo.rb`
 
```
Offenses:

foo.rb:1:9: C: [Corrected] Layout/SpaceAfterComma: Space missing after comma.
foo { |a,; x|
        ^
foo.rb:1:10: C: [Corrected] Layout/SpaceBeforeSemicolon: Space found before semicolon.
foo { |a, ; x|
         ^

0 files inspected, 2 offenses detected, 2 offenses corrected
Infinite loop detected in foo.rb and caused by Layout/SpaceBeforeSemicolon -> Layout/SpaceAfterComma
/Users/snutij/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-1.54.0/lib/rubocop/runner.rb:334:in `check_for_infinite_loop'
/Users/snutij/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-1.54.0/lib/rubocop/runner.rb:317:in `block in iterate_until_no_changes'
/Users/snutij/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-1.54.0/lib/rubocop/runner.rb:316:in `loop'
/Users/snutij/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-1.54.0/lib/rubocop/runner.rb:316:in `iterate_until_no_changes'
/Users/snutij/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-1.54.0/lib/rubocop/runner.rb:285:in `do_inspection_loop'
/Users/snutij/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-1.54.0/lib/rubocop/runner.rb:165:in `block in file_offenses'
/Users/snutij/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-1.54.0/lib/rubocop/runner.rb:190:in `file_offense_cache'
/Users/snutij/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-1.54.0/lib/rubocop/runner.rb:164:in `file_offenses'
/Users/snutij/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-1.54.0/lib/rubocop/runner.rb:155:in `process_file'
/Users/snutij/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-1.54.0/lib/rubocop/runner.rb:135:in `block in each_inspected_file'
/Users/snutij/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-1.54.0/lib/rubocop/runner.rb:134:in `each'
/Users/snutij/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-1.54.0/lib/rubocop/runner.rb:134:in `reduce'
/Users/snutij/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-1.54.0/lib/rubocop/runner.rb:134:in `each_inspected_file'
/Users/snutij/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-1.54.0/lib/rubocop/runner.rb:120:in `inspect_files'
/Users/snutij/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-1.54.0/lib/rubocop/runner.rb:73:in `run'
```

## Steps to reproduce the problem

In `RuboCop::Runner` the method `check_for_infinite_loop` raises the error, 

The if condition is defined by:

```ruby 
(ruby) loop_start_index = @processed_sources.index(checksum)
2
```

I can provide logs for `processed_source` & `offenses_by_iteration` but it is very large data, but AST can be useful right now: 

```ruby
(ruby) processed_source.ast
s(:block,
  s(:send, nil, :foo),
  s(:args,
    s(:arg, :a),
    s(:shadowarg, :x)), nil)
```

https://github.com/rubocop/rubocop/blob/0932941b791812e91394d03a4cf869517fc63983/lib/rubocop/runner.rb#L328-L340

## RuboCop version

```
bundle exec rubocop -V

1.54.0 (using Parser 3.2.2.3, rubocop-ast 1.29.0, running on ruby 3.2.2) [arm64-darwin22]
  - rubocop-minitest 0.31.0
  - rubocop-rake 0.6.0
  - rubocop-sorbet 0.7.0
```
